test/test_scratchpad.py

```diff
@@ -71,6 +71,7 @@ class ScratchPadBaseConfig(Config):
     floating_layout = libqtile.resources.default_config.floating_layout
     keys = []
     mouse = []
+    reconfigure_screens = False
```

Comment:
While this makes the test pass, this can't be the solution. However, I just don't understand why this PR stops the `client_focus` hook from firing here.

Bizarrely, `test/test_hook.py::test_client_focus` passes, even when `reconfigure_screens = True`, so there's something else going on too.

Reply:
Yeah, absolutely agreed. Sorry, I didn't intend to merge this, I was just experimenting. I think I do understand what is going on here, but I couldn't manage to "fix" it with sleeps, so I'm not completely convinced.

We are getting a ScreenChangeNotify event (not the fake one I added, but a real one from the X11 server) during *every* test now with this change. I found this doc: https://www.x.org/archive/X11R6.8.1/doc/Xrandr.3.html#toc4 which says:

> Additionally, to eliminate a potential race condition, this event may be generated immediately upon selecting for notification if the screen has changed since the client of Xrandr connected to the X server, to enable reliable screen resolution changing when a user may log in and change the configuration while one or many clients are starting up.

So my theory is: we are unmasking ScreenChangeNotify much later than previously, right before we hand off control to the test suite. So this means that the test suite sees this ScreenChangeNotify, which it previously did not. Most tests must be agnostic to this, but it seems this particular Scratchpad one is not? If that's the case, you'd think you could put a `time.sleep(5)` at the beginning of the failing test and get it to pass, but that didn't work for me. I tried flushing the X11 connection and a few other things, no luck.

Another option I had considered was that something about the scratchpad code was generating this ScreenChangeNotify, but I didn't see anything suspicious there. Plus, given that it happens on *other* tests as well (e.g. I tried with the dgroups ones and see the same behavior), I suspect it's unrelated to the actual scratchpad test.

The final piece of weirdness in all of this: I explicitly disable Xephyr's randr here, so in principle we shouldn't be getting any ScreenChangeNotify events at all, because the randr extension shouldn't exist. But I think Xephyr's implementation of randr is pretty poor based on my previous experimentation, so I haven't read a lot into this.

We could try and switch all of the testing to the x11's "dummy" driver instead of these xvfb/xephyr hacks, which would enable us to actually test my randr stuff in this branch. I'm still trying to validate my theory, hopefully I can get there by adding some extra synchronization points to the tests.

Reply:
So, is the scratchpad test failing because it's the only test that has `reconfigure_screens = True`? I find it a bit concerning that that apparently stops the hook being fired.

Reply:
> the hook being fired.

Here you meant the `client_new` hook, not `screen_changed`? I agree it is weird and shouldn't do that, there is more weirdness going on on our side. But I suspect that weirdness has always been there, and it is only the moving of the ScreenChangeNotify that is triggering it.

Reply:
I am still looking at this, and have no real leads.

I did get the xtrace stuff working, though, which has been helpful to understand what's going on. Seems like maybe I should pull that and the other spelling fix out to a separate branch to land them in case we need this in the future.